### PR TITLE
NIFI-5895 - Add a processor to start an Oozie workflow

### PR DESCRIPTION
**Pull request by pvillard31**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by pvillard31** [comment]
This PR has been tested against an Oozie 4.2.0 server instance in the following conditions:
- Plaintext
- Plaintext + SSL
- Kerberos
- Kerberos + SSL

A very basic examples has been used:
https://github.com/apache/oozie/tree/branch-4.3/examples/src/main/apps/shell

Testing workflow:
https://gist.github.com/pvillard31/4a0a7b33dc27c11538cec0b5e80c0324

The workflow.xml from the Oozie repo must be pushed in HDFS in a directory and that directory is used in the configuration of the processor. All the properties given in the job.properties file from the Oozie repo must be passed through the configuration of the processor or the attributes of the flow files.

File: nifi-nar-bundles/nifi-oozie-bundle/nifi-oozie-processors/src/main/java/org/apache/nifi/processors/oozie/SubmitOozieWorkflow.java [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+            .displayName("Oozie server URL")
+            .description("The URL of the Oozie server. Example: http://oozie.example.com:11000/oozie")
+            .required(true)
+            .addValidator(StandardValidators.URL_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .build();
+
+    public static final PropertyDescriptor APP_PATH = new PropertyDescriptor.Builder()
+            .name("oozie-app-path")
+            .displayName("HDFS Application Path")
```
Comment (zenfenan) [inline nifi-nar-bundles/nifi-oozie-bundle/nifi-oozie-processors/src/main/java/org/apache/nifi/processors/oozie/SubmitOozieWorkflow.java]: I would recommend it to be changed to something like `Workflow Path`.

File: nifi-nar-bundles/nifi-oozie-bundle/nifi-oozie-processors/pom.xml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        <groupId>org.apache.nifi</groupId>
+        <artifactId>nifi-oozie-bundle</artifactId>
+        <version>1.9.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>nifi-oozie-processors</artifactId>
+    <packaging>jar</packaging>
+
+    <dependencies>
+		<dependency>
```
Comment (zenfenan) [inline nifi-nar-bundles/nifi-oozie-bundle/nifi-oozie-processors/pom.xml]: Pure nitpick: Would like to see proper indentation. :)

**Comment by pvillard31** [comment]
Thanks for the review @zenfenan - just pushed a commit to address your comments.

File: nifi-nar-bundles/nifi-oozie-bundle/nifi-oozie-processors/src/main/java/org/apache/nifi/processors/oozie/SubmitOozieWorkflow.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+            return;
+        }
+
+        final KerberosCredentialsService credentialsService = context.getProperty(KERBEROS_CREDENTIALS_SERVICE).asControllerService(KerberosCredentialsService.class);
+        if(credentialsService != null) {
+            try {
+                String principal = credentialsService.getPrincipal();
+                String keyTab = credentialsService.getKeytab();
+                Configuration conf = new Configuration();
+                conf.set("hadoop.security.authentication", "kerberos");
```
[the remark this answers is not in the document]
Comment (pvillard31) [reply nifi-nar-bundles/nifi-oozie-bundle/nifi-oozie-processors/src/main/java/org/apache/nifi/processors/oozie/SubmitOozieWorkflow.java]: Not to my knowledge - IIRC this Kerberos property is to enable SPNEGO authN against the Oozie server REST API but there should not be any other requirement. The files hdfs-site, core-site, etc, are loaded by the Oozie server anyway.

File: nifi-nar-bundles/nifi-oozie-bundle/nifi-oozie-nar/src/main/resources/META-INF/NOTICE [hunk 4]
```diff
@@ -0,0 +1,136 @@
+nifi-kafka-nar
```
[the remark this answers is not in the document]
Comment (pvillard31) [reply nifi-nar-bundles/nifi-oozie-bundle/nifi-oozie-nar/src/main/resources/META-INF/NOTICE]: Good catch - I tried to be exhaustive on the Oozie deps, hopefully didn't miss one.

File: nifi-nar-bundles/nifi-oozie-bundle/pom.xml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    </parent>
+
+    <groupId>org.apache.nifi</groupId>
+    <artifactId>nifi-oozie-bundle</artifactId>
+    <version>1.9.0-SNAPSHOT</version>
+    <packaging>pom</packaging>
+    
+    <properties>
+    	<oozie.version>4.3.1</oozie.version>
+    	<hadoop.version>3.0.0</hadoop.version>
```
[the remark this answers is not in the document]
Comment (pvillard31) [reply nifi-nar-bundles/nifi-oozie-bundle/pom.xml]: Yep true. Didn't look too long (switching planes in airport) but didn't find a way to inherit the version from the nifi-hadoop-libraries so I just added a comment about the version.

**Comment by joewitt** [comment]
closing due to inactivity

**Pull request closed without merging**